### PR TITLE
Engine auditd assets

### PR DESCRIPTION
**Pull request by nmkoremblum**
|Related issues|
|---|
|#13870|

As described in issue #13870, it is required to create assets to process `auditd` events.

To do so, we need to create a decoder that generates an ECS-compliant events, and rules to alert about possible security threats.

We need to use:

- high-level parser
- logical expressions
- function helpers
- mapping new fields
- kvdb

All the assets will become part of a single environment that will receive logs and generate the corresponding events.

### CODE REVIEW

**Review by juliancnn — APPROVED** [review]
LGTM!

File: src/engine/source/hlp/tools/example.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 -
-  logql_expression: "<_map/MAP/ /=/.> <_dummy>"
+  logql_expression: "<_map/kv_map/=/ />. <_dummy>"
   event: "key1=Value1 Key2=Value2. dummy"
 -
-  logql_expression: "<_map/MAP/ /=>"
+  logql_expression: "<_map/kv_map/=/ >"
   event: "key1=Value1 Key2="
```
Comment (JavierBejMen) [inline src/engine/source/hlp/tools/example.yml:6]: Why is this file added here?

File: src/engine/test/assets/decoders/fim.yml [hunk 2]
```diff
@@ -1,3 +1,4 @@
+---
```
Comment (JavierBejMen) [inline src/engine/test/assets/decoders/fim.yml:1]: This line is only needed if multiple Yaml documents are defined on the same file, don't hurts though.

File: src/engine/test/source/hlp/CMakeLists.txt [hunk 3]
```diff
@@ -6,3 +6,12 @@ add_executable(hlp_test
 target_link_libraries(hlp_test hlp gtest_main)
 
 gtest_discover_tests(hlp_test)
+
+# Configuration Catalog tests
```
Comment (JavierBejMen) [inline src/engine/test/source/hlp/CMakeLists.txt:10]:
```suggestion
# Configuration KVMap tests
```

**Review by JavierBejMen — CHANGES_REQUESTED** [review]

File: src/engine/test/assets/decoders/auditd.yml [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      - _log.msg: +exists
+    map:
+      _kv_msg: $_log.msg
+
+  - map:
+      event.kind: event
+      timestamp: $_log.epoch
+      event.action: +s_lo/$_log.record_type
+      wazuh.avc: +rename_field/$_avc
+      _log: +delete_field
```
Comment (LucioDonda) [inline src/engine/test/assets/decoders/auditd.yml:153]: didn't we decide to leave temporary variables without erasing them? (when correctly implemented should be automatically deleted at the end of the asset)

File: src/engine/test/source/builder/builders/opBuilderHelperRenameField_test.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+using namespace base;
+using namespace builder::internals::builders;
+
+using std::string;
+using std::vector;
+
+base::EngineOp op {};
+
+TEST(opBuilderHelperRenameField, build)
```
Comment (LucioDonda) [inline src/engine/test/source/builder/builders/opBuilderHelperRenameField_test.cpp:17]: what if the source field is an array, or a nested object, or it has a null or empty value? are we checking those cases?

File: src/engine/source/hlp/src/specificParsers.cpp [hunk 6]
```diff
@@ -323,74 +332,281 @@ bool parseJson(const char** it, Parser const& parser, ParseResult& result)
...
+    {
+        rapidjson::StringBuffer s;
+        rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+        output_doc.Accept(writer);
+        result[parser.name] = hlp::JsonString {s.GetString()};
+        *it = lastFoundOk;
+    }
 
-    return true;
+    return lastFoundOk == nullptr ? false : true;
```
Comment (LucioDonda) [inline src/engine/source/hlp/src/specificParsers.cpp:609]: this should be the same, right?
`return lastFoundOk == nullptr`

File: src/engine/source/hlp/src/specificParsers.cpp [hunk 7]
```diff
@@ -323,74 +332,281 @@ bool parseJson(const char** it, Parser const& parser, ParseResult& result)
...
+            {
+                // No endMapToken, no pairSeparator, no end quoteChar
+                break;
+            }
         }
-        tuple_start_pos = tuple_end_pos + 1;
+        // Print key and value and iterator
+        lastFoundOk = strParsePtr;
+        // std::cout << "Key: '" << key << "', Value: '" << value << "', Iterator: '" <<
+        // strParsePtr << "'" << std::endl;
```
Comment (LucioDonda) [inline src/engine/source/hlp/src/specificParsers.cpp:552]: erase comment

File: src/engine/test/source/hlp/hlp_test.cpp [hunk 8]
```diff
@@ -349,13 +352,64 @@ TEST(hlpTests_map, incomplete_map_test)
...
+    const char* logQl = "<_map/kv_map/: / > hi!";
+    const char* event = "key1: Value1 Key2: Value2 hi!";
+
+    ParserFn parseOp = getParserOp(logQl);
+    ASSERT_EQ(true, static_cast<bool>(parseOp));
+
+    ParseResult result;
+    bool ret = parseOp(event, result);
+
+    ASSERT_TRUE(static_cast<bool>(parseOp));
```
Comment (LucioDonda) [inline src/engine/test/source/hlp/hlp_test.cpp:404]: why are we testing some things here and others in hlpKVMap_test ?

File: src/engine/source/hlp/tools/example.yml (hunk 1 above)
Comment (juliancnn) [reply src/engine/source/hlp/tools/example.yml:6]: It is the hlp tool for testing parsers. Updated parser name.

File: src/engine/test/source/hlp/CMakeLists.txt (hunk 3 above)
Comment (juliancnn) [reply src/engine/test/source/hlp/CMakeLists.txt:10]: good catch

File: src/engine/test/source/hlp/hlp_test.cpp (hunk 8 above)
Comment (juliancnn) [reply src/engine/test/source/hlp/hlp_test.cpp:612]: It's true, this shouldn't be here

**Review by LucioDonda — APPROVED** [review]
LGTM

**Review by JcabreraC — APPROVED** [review]
LGTM !
